maven/new_rules.bzl

```diff
@@ -0,0 +1,268 @@
+        outputs = [output_jar],
+        arguments = [
+            "--prefix=",  # prefix is deliberately left empty
+            "--group-id=" + maven_coordinates.group_id,
+            "--artifact-id=" + maven_coordinates.artifact_id,
+            "--pom-file=" + pom_file.path,
+            "--jars=" + ";".join(class_jar_paths),
+            "--output=" + output_jar.path,
+        ],
+    )
+
+    return output_jar
+
+def _generate_source_jar(ctx):
+    target = ctx.attr.target
+    maven_coordinates = _parse_maven_coordinates(target[AggregatedDependencyInfo].maven_coordinates)
+
+    srcjar = None
+
+    if hasattr(target, "files") and target.files.to_list() and target.files.to_list()[0].extension == "jar":
+        for output in target[JavaInfo].outputs.jars:
+            if output.source_jar and (output.source_jar.basename.endswith("-src.jar") or output.source_jar.basename.endswith("-sources.jar")):
+                srcjar = output.source_jar
+                break
+    else:
+        fail("Could not find JAR file to deploy in {}".format(target))
+
+    if not srcjar:
+        return None
+
+    output_jar = ctx.actions.declare_file("{}:{}-sources.jar".format(maven_coordinates.group_id, maven_coordinates.artifact_id))
+
+    source_jar_deps = [dep.source_jar for dep in target[AggregatedDependencyInfo].deps.to_list() if dep.type == 'jar']
+    source_jar_paths = [srcjar.path] + [target.path for target in source_jar_deps]
+
+    ctx.actions.run(
+        executable = ctx.executable._jar_assembler,
+        inputs = [srcjar] + source_jar_deps,
+        outputs = [output_jar],
+        arguments = [
+            "--prefix=" + ctx.attr.source_jar_prefix,
+            "--jars=" + ";".join(source_jar_paths),
+            "--output=" + output_jar.path,
+        ],
+    )
+
+    return output_jar
+
+def _assemble_maven_impl(ctx):
+    print(ctx.attr.target[AggregatedDependencyInfo])
+    version_file = _construct_version_file(ctx)
+    pom_file = _generate_pom_file(ctx, version_file)
+    class_jar = _generate_class_jar(ctx, pom_file)
+    source_jar = _generate_source_jar(ctx)
+
+    output_files = [pom_file, class_jar]
+    if source_jar:
+        output_files.append(source_jar)
+
+    return [
+        DefaultInfo(files = depset(output_files)),
+        MavenDeploymentInfo(jar = class_jar, pom = pom_file, srcjar = source_jar)
+    ]
+
+def find_maven_coordinates(target, tags):
+    _TAG_KEY_MAVEN_COORDINATES = "maven_coordinates="
+    for tag in tags:
+        if tag.startswith(_TAG_KEY_MAVEN_COORDINATES):
+            coordinates = tag[len(_TAG_KEY_MAVEN_COORDINATES):]
+            target_is_in_root_workspace = target.label.workspace_root == ""
+            if coordinates.endswith("{pom_version}") and not target_is_in_root_workspace:
+                coordinates = coordinates.replace("{pom_version}", target.label.workspace_root.replace("external/", ""))
+            return coordinates
+
+AggregatedDependencyInfo = provider(
```

Comment:
this is looking really good now! the new definition is really easy to understand and i think can be renamed:

```
JarInfo = provider(
    fields = {
        "name": "The name of a the JAR (Maven coordinates)",
        "deps": "The dependency of this JAR. It may be of two types, POM or JAR.",
    },
)
```